### PR TITLE
Added the ability to access Spec and parameters as JSON from Python

### PR DESCRIPTION
**Pull request by dkeeney**
New function Network.getSpecJSON(region_type) will retrieve the spec for the specified region type.  The region does not have to be created first.

New function Region.getParameters() will retrieve all parameters on a region as a single JSON string.
New function Region.getParameterJSON( ) will retrieve the value of a specific parameter as a JSON string.

New unit test for python is in py/tests/networkapi/sinewave_test.py
  This is a python version of the simple sinewave example.

New C++ unit tests were added to some of the region implementation tests to confirm the above functions.

This is a specific request from @Zbysekz

### CODE REVIEW

**Comment by Zbysekz** [comment]
Firstly, thank you for this 👍 

I tried it directly with advanced regions implemented in python, but i got following

 `E       RuntimeError: Exception: Region.cpp(631) message: Error setting parameter py.ColumnPoolerRegion.learningTolerance; AttributeError: 'ColumnPoolerRegion' object has no attribute 'learningTolerance'
E       `
See commit above for reproducing this (i have just made rough modification of sinewave_test).

I guess it is about to align custom python regions to support this, can you guide me please?

**Comment by Zbysekz** [comment]
Agh, it just revealed that learningTolerance is only in spec but not used anywhere.. so bug in py.ColumnPoolerRegion.
Removing it from spec moves me further:
now i am getting:
```
print(region[1].getParameters())
RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)
```

**Comment by dkeeney** [comment]
> json = encoder.getParameters()
E       RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)

This is odd.  
Ok, let me get your changes and I will see if I can figure out what is happening.

By the way.  I do OK with C++ but I do not have a lot of Python experience so I appreciate your help in finding these things.

**Comment by dkeeney** [comment]
The values for two of the parameters were None.  Or they could be the default_value from the spec which is "".
```
         maxSdrSize = None,
         minSdrSize = None,
```
This is a valid value.  But if the C++ data type is numeric then casting caused the error.
So I first checked to see if the value is None and if it is I return as cast of "0".

Now the error is "RuntimeError: Could not allocate capsule object!"
I don't know what that means so I have to research it.  Something to do with pybind11 I think.

**Comment by Zbysekz** [comment]
I tried py.ApicalTMPairRegion with error, then py.RawSensor and that was working without issue, so one difference hits my eyes:
There is `defaultOutputType `parameter, which is Byte type and it is enumeration
```
defaultOutputType=dict(
                    description="Controls what type of cell output is placed into"
                                " the default output 'feedForwardOutput'",
                    accessMode="Create",
                    dataType="Byte",
                    count=0,
                    constraints="enum: active,predicted,predictedActiveCells",
                    defaultValue="active"),

```

And this is causing the ` "RuntimeError: Could not allocate capsule object!"`

**Comment by Zbysekz** [comment]
More info:

- there is next bug in GridCellLocationRegion.py : maxSynapsesPerSegment parameter is UInt32 but default value is "-1", leads to  `RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)` this is easy to correct
- scale and orientation in GridCellLocationRegion.py are real32, but what is unique, is that they are arrays. They are expected as arrays of size "moduleCount". Leads to `"RuntimeError: Could not allocate capsule object!"`

I have removed in my draft all these problematic params and it's passing..

So to summarize problem, i think we need to handle these types of parameters:

- [ ] enumeration
- [ ] arrays
- [ ] if it's possible maybe some better exception throw message to find what parameter is causing it?

**Comment by dkeeney** [comment]
Thanks for hunting this down.
- a default value of -1 for a UInt is just an error in the Spec.  So either the default value or the type must be changed.
- You are right, we do not handle enumerations. The were just not implemented.   The problem is that we should implement enumerations both in C++ and Python to be consistent and  C++ requires that the enumerations be available at compile time. I am inclined to change the parameter to an integer and avoid the whole issue.  Let ApicalTMPairRegion map the integer to enumerations.
- On the C++ side we do handle arrays but we might not he doing it right in the Python bindings.  I think I know where that might be.
- And the error message definitely needs some attention. I think it is generated by pybind11 or perhaps Python itself in the bindings.

Enumerations would require some new design work to implement.  It is probably not too hard to find a way to map integers to a list of names in Python.  But for C++, the enum names must exist at compile time.  Since we have just this one instance I am inclined to replace it with an integer and having the

**Comment by dkeeney** [comment]
Oh, looking at ApicalTMPairRegion  and the implementation field, its actually a String.
So change the type to "String" and the count to 1.   That should fix it.

**Comment by breznak** [comment]
> So change the type to "String" and the count to 1. That should fix it. 

looking to review now, does this answer all problems in https://github.com/htm-community/htm.core/pull/866#issuecomment-668579268 , or only the ApicalTM stuff? 

@Zbysekz do you want to add a unit test with your crashing conditions?

File: src/htm/engine/Network.hpp [hunk 1]
```diff
@@ -117,6 +117,7 @@ class Link;
    *  On errors it throws an exception.
    */
   void configure(const std::string &yaml);
+  std::string getSpecJSON(const std::string &type);
```
Comment (breznak) [inline src/htm/engine/Network.hpp:120]: please add a short doc on the functionality,usage. and what the type param should be.

File: src/htm/engine/Region.hpp [hunk 2]
```diff
@@ -143,7 +139,20 @@ class Region : public Serializable {
...
+   * @param name
+   *        The name of the parameter
+   *
+   * @param tag  (optional)
+   *       if the tag IS NOT given (or blank), it returns just the JSON encoded value.
+   *       if the tag IS given, it returns <tag>: {"value": <JSON value>, "type": <type>}
+   *
+   * @returns The value of the parameter
+   */
+  std::string getParameterJSON(const std::string &name, bool withType = false) const;
```
Comment (breznak) [inline src/htm/engine/Region.hpp:155]: ok, the doc is here..

File: src/htm/regions/TestNode.cpp [hunk 3]
```diff
@@ -216,6 +216,7 @@ void TestNode::compute() {
 Spec *TestNode::createSpec() {
   auto ns = new Spec;
 
+  ns->name = "TestNode";
   ns->description = "TestNode. Used as a plain simple plugin Region for unit tests only. "
```
Comment (breznak) [inline src/htm/regions/TestNode.cpp:220]: nit, do we need the TestNode , or could we simply use any of the SP,TMRegions?

Comment (dkeeney) [reply src/htm/regions/TestNode.cpp:220]: Many of the unit tests use this as a "generic" region.  It contains a little of everything.  Not all regions actually use all features.

**Comment by dkeeney** [comment]
This is not ready for merge yet.  
We found that some parts were not working quite right when applied to regions implemented in Python.
I should have another push later today.

**Comment by Zbysekz** [comment]
> * a default value of -1 for a UInt is just an error in the Spec.  So either the default value or the type must be changed.

Yes please change just datatype from UInt32 to Int32 and should be fine

> * You are right, we do not handle enumerations. The were just not implemented.   The problem is that we should implement enumerations both in C++ and Python to be consistent and  C++ requires that the enumerations be available at compile time. I am inclined to change the parameter to an integer and avoid the whole issue.  Let ApicalTMPairRegion map the integer to enumerations.

Yes i agree

> * And the error message definitely needs some attention. I think it is generated by pybind11 or perhaps Python itself in the bindings.

I will try to look at that, maybe tomorrow will have some spare time

**Comment by dkeeney** [comment]
Ok, I think I have getParameters( ) working correctly now.

There are a lot of changes because I found that all of the parameter getters (like getParameterUInt32(), etc.) for all C++ regions were not const functions.  The Region class had them as const but the implementations which it called were not.  Do not know how the compilers let me get away with that.  Anyway, it is consistent now and all getters are const.  I just hope I did not break anyone.
